txnsync/txngroups.go

```diff
@@ -17,23 +17,417 @@
+type assetIndeces []basics.AssetIndex
+
+//msgp:allocbound program config.MaxAppProgramLen
+type program []byte
+
+type encodedApplicationCallTxnFields struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+	ApplicationID        []basics.AppIndex `codec:"apid,allocbound=maxEncodedTransactionGroup"`
+	BitmaskApplicationID bitmask           `codec:"apidbm"`
+
+	OnCompletion        []transactions.OnCompletion `codec:"apan,allocbound=maxEncodedTransactionGroup"`
+	BitmaskOnCompletion bitmask                     `codec:"apanbm"`
+
+	ApplicationArgs        []applicationArgs `codec:"apaa,allocbound=maxEncodedTransactionGroup"`
+	BitmaskApplicationArgs bitmask           `codec:"apaabm"`
+
+	Accounts        []addresses `codec:"apat,allocbound=maxEncodedTransactionGroup"`
+	BitmaskAccounts bitmask     `codec:"apatbm"`
+
+	ForeignApps        []appIndeces `codec:"apfa,allocbound=maxEncodedTransactionGroup"`
+	BitmaskForeignApps bitmask      `codec:"apfabm"`
+
+	ForeignAssets        []assetIndeces `codec:"apas,allocbound=maxEncodedTransactionGroup"`
+	BitmaskForeignAssets bitmask        `codec:"apasbm"`
+
+	LocalStateSchema        []basics.StateSchema `codec:"apls,allocbound=maxEncodedTransactionGroup"`
+	BitmaskLocalStateSchema bitmask              `codec:"aplsbm"`
+
+	GlobalStateSchema        []basics.StateSchema `codec:"apgs,allocbound=maxEncodedTransactionGroup"`
+	BitmaskGlobalStateSchema bitmask              `codec:"apgsbm"`
+
+	ApprovalProgram        []program `codec:"apap,allocbound=maxEncodedTransactionGroup"`
+	BitmaskApprovalProgram bitmask   `codec:"apapbm"`
+
+	ClearStateProgram        []program `codec:"apsu,allocbound=maxEncodedTransactionGroup"`
+	BitmaskClearStateProgram bitmask   `codec:"apsubm"`
+}
+
+type encodedCompactCertTxnFields struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+	CertRound        []basics.Round `codec:"certrnd,allocbound=maxEncodedTransactionGroup"`
+	BitmaskCertRound bitmask        `codec:"certrndbm"`
+
+	CertType        []protocol.CompactCertType `codec:"certtype,allocbound=maxEncodedTransactionGroup"`
+	BitmaskCertType bitmask                    `codec:"certtypebm"`
+
+	Cert        []compactcert.Cert `codec:"cert,allocbound=maxEncodedTransactionGroup"`
+	BitmaskCert bitmask            `codec:"certbm"`
+}
+
+const (
+	paymentTx = iota
+	keyRegistrationTx
+	assetConfigTx
+	assetTransferTx
+	assetFreezeTx
+	applicationCallTx
+	compactCertTx
+	unknownTx
+)
+
+// TxTypeToByte converts a TxType to byte encoding
+func TxTypeToByte(t protocol.TxType) byte {
+	switch t {
+	case protocol.PaymentTx:
+		return paymentTx
+	case protocol.KeyRegistrationTx:
+		return keyRegistrationTx
+	case protocol.AssetConfigTx:
+		return assetConfigTx
+	case protocol.AssetTransferTx:
+		return assetTransferTx
+	case protocol.AssetFreezeTx:
+		return assetFreezeTx
+	case protocol.ApplicationCallTx:
+		return applicationCallTx
+	case protocol.CompactCertTx:
+		return compactCertTx
+	default:
+		return unknownTx
```

Comment:
There is no `unknownTx`. We shouldn't generate a value for that. instead, if we get it here, we should exit with an error.